src/cascade/dismod/db/wrapper.py

```diff
@@ -0,0 +1,153 @@
+    return engine
+
+
+class DismodFile:
+    """
+    Responsible for creation of a Dismod-AT file.
+
+    This class checks the type of all columns. It doesn't check that the
+    model is correct::
+
+        engine = _get_engine(None)
+        dm = DismodFile(engine, {"col": float}, {})
+        dm_file.add("time", pd.DataFrame({"time": [1997, 2005, 2017]}))
+        time_df = dm_file.time
+
+    The arguments for ``avgint_columns`` and ``data_columns`` add columns
+    to the avgint and data tables. These arguments are dictionaries from
+    column name to column type.
+    """
+
+    def __init__(self, engine, avgint_columns, data_columns):
+        """
+        The columns arguments add columns to the avgint and data
+        tables.
+
+        Args:
+            engine: A sqlalchemy engine
+            avgint_columns (dict): From columns to types.
+            data_columns (dict): From columns to types.
+        """
+        self.engine = engine
+        self._table_definitions = Base.metadata.tables
+        self._table_data = {}
+        self._table_hash = {}
+        add_columns_to_avgint_table(avgint_columns)
+        add_columns_to_data_table(data_columns)
+        LOGGER.debug(f"dmfile tables {self._table_definitions.keys()}")
+
+    def create_tables(self):
+        """
+        Make all of the tables in the metadata.
+        """
+        # TODO: we only actually need to make a subset of the tables
+        Base.metadata.create_all(self.engine)
+
+    def make_densities(self):
+        """
+        Dismod documentation says all densities should be in the file,
+        so this puts them all in.
+        """
+        self.density = pd.DataFrame({"density": [x.name for x in DensityEnum]})
+
+    def __getattr__(self, table_name):
+        if table_name in self._table_data:
+            return self._table_data[table_name]
+        elif table_name in self._table_definitions:
+            table = self._table_definitions[table_name]
+            with self.engine.connect() as conn:
+                data = pd.read_sql_query(select([table]), conn)
+            data = data.set_index(f"{table_name}_id")
+            self._table_hash[table_name] = pd.util.hash_pandas_object(data)
+            self._table_data[table_name] = data
+            return data
+        else:
+            raise AttributeError(f"No such table {table_name}")
+
+    def __setattr__(self, table_name, df):
+        if table_name in self.__dict__.get("_table_definitions", {}):
+            self._table_data[table_name] = df
+        else:
+            super().__setattr__(table_name, df)
+
+    def _is_dirty(self, table_name):
+        """Tests to see if the table's data has changed in memory since it was last loaded from the database.
+        """
+
+        if table_name not in self._table_data:
+            return False
+
+        table = self._table_data[table_name]
+        table_hash = pd.util.hash_pandas_object(table)
```

Comment:
Meh, I'm not super happy with this. But the fact that the DataFrames are mutable it's tricky to track whether they are dirty without writing a complex proxy on top of them. This is the CPU and memory intensive way of avoiding that code complexity. But I doubt it will cause any trouble in practice.